From cff648be60b8cc6c74343f4cc75ed99ba9245372 Mon Sep 17 00:00:00 2001
From: Guojing Yuan <guojing@google.com>
Date: Tue, 6 May 2025 21:36:19 +0000
Subject: [PATCH 01/56] Revoke NLS when all associations are removed.

Bug: 391895151
Test: manual test
Flag: EXEMPT bugfix
(cherry picked from commit a987e9d91a93b5945ac7d6477e6d34cd2715f2ec)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:f243250a5a1687615f55447de3dca31b397821f1
Merged-In: I09ac1ef07b65e5180d38e8697d4507a113162eb7
Change-Id: I09ac1ef07b65e5180d38e8697d4507a113162eb7
---
 .../CompanionDeviceManagerService.java        |  5 +++-
 .../association/DisassociationProcessor.java  | 29 +++++++++++++++++--
 2 files changed, 31 insertions(+), 3 deletions(-)

diff --git a/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java b/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
index 76f153ad4..7197c3414 100644
--- a/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
+++ b/services/companion/java/com/android/server/companion/CompanionDeviceManagerService.java
@@ -155,6 +155,8 @@ public class CompanionDeviceManagerService extends SystemService {
         final UserManager userManager = context.getSystemService(UserManager.class);
         final PowerManagerInternal powerManagerInternal = LocalServices.getService(
                 PowerManagerInternal.class);
+        final NotificationManager notificationManager = context.getSystemService(
+                NotificationManager.class);
 
         final AssociationDiskStore associationDiskStore = new AssociationDiskStore();
         mAssociationStore = new AssociationStore(context, userManager, associationDiskStore);
@@ -182,7 +184,8 @@ public class CompanionDeviceManagerService extends SystemService {
 
         mDisassociationProcessor = new DisassociationProcessor(context, activityManager,
                 mAssociationStore, packageManagerInternal, mDevicePresenceProcessor,
-                mCompanionAppBinder, mSystemDataTransferRequestStore, mTransportManager);
+                mCompanionAppBinder, mSystemDataTransferRequestStore, mTransportManager,
+                notificationManager);
 
         mSystemDataTransferProcessor = new SystemDataTransferProcessor(this,
                 packageManagerInternal, mAssociationStore,
diff --git a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
index ea5ebc720..928a08afb 100644
--- a/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
+++ b/services/companion/java/com/android/server/companion/association/DisassociationProcessor.java
@@ -28,13 +28,17 @@ import android.annotation.NonNull;
 import android.annotation.SuppressLint;
 import android.annotation.UserIdInt;
 import android.app.ActivityManager;
+import android.app.NotificationManager;
 import android.companion.AssociationInfo;
 import android.companion.Flags;
 import android.content.Context;
+import android.content.Intent;
 import android.content.pm.PackageManagerInternal;
+import android.content.pm.ResolveInfo;
 import android.os.Binder;
 import android.os.SystemProperties;
 import android.os.UserHandle;
+import android.service.notification.NotificationListenerService;
 import android.util.Slog;
 
 import com.android.server.companion.datatransfer.SystemDataTransferRequestStore;
@@ -42,6 +46,8 @@ import com.android.server.companion.devicepresence.CompanionAppBinder;
 import com.android.server.companion.devicepresence.DevicePresenceProcessor;
 import com.android.server.companion.transport.CompanionTransportManager;
 
+import java.util.List;
+
 /**
  * This class responsible for disassociation.
  */
@@ -77,6 +83,7 @@ public class DisassociationProcessor {
     private final CompanionTransportManager mTransportManager;
     private final OnPackageVisibilityChangeListener mOnPackageVisibilityChangeListener;
     private final ActivityManager mActivityManager;
+    private final NotificationManager mNotificationManager;
 
     public DisassociationProcessor(@NonNull Context context,
             @NonNull ActivityManager activityManager,
@@ -85,7 +92,8 @@ public class DisassociationProcessor {
             @NonNull DevicePresenceProcessor devicePresenceMonitor,
             @NonNull CompanionAppBinder applicationController,
             @NonNull SystemDataTransferRequestStore systemDataTransferRequestStore,
-            @NonNull CompanionTransportManager companionTransportManager) {
+            @NonNull CompanionTransportManager companionTransportManager,
+            @NonNull NotificationManager notificationManager) {
         mContext = context;
         mActivityManager = activityManager;
         mAssociationStore = associationStore;
@@ -96,12 +104,12 @@ public class DisassociationProcessor {
         mCompanionAppController = applicationController;
         mSystemDataTransferRequestStore = systemDataTransferRequestStore;
         mTransportManager = companionTransportManager;
+        mNotificationManager = notificationManager;
     }
 
     /**
      * Disassociate an association by id.
      */
-    // TODO: also revoke notification access
     public void disassociate(int id, String reason) {
         Slog.i(TAG, "Disassociating id=[" + id + "]...");
 
@@ -136,6 +144,23 @@ public class DisassociationProcessor {
         mSystemDataTransferRequestStore.removeRequestsByAssociationId(userId, id);
         mAssociationStore.removeAssociation(association.getId(), reason);
 
+        // Revoke NLS if the last association has been removed for the package
+        Binder.withCleanCallingIdentity(() -> {
+            if (mAssociationStore.getAssociationsByPackage(userId, packageName).isEmpty()) {
+                Intent nlsIntent = new Intent(
+                        NotificationListenerService.SERVICE_INTERFACE);
+                List<ResolveInfo> matchedServiceList = mContext.getPackageManager()
+                        .queryIntentServicesAsUser(nlsIntent, /* flags */ 0, userId);
+                for (ResolveInfo service : matchedServiceList) {
+                    if (service.getComponentInfo().getComponentName().getPackageName()
+                            .equals(packageName)) {
+                        mNotificationManager.setNotificationListenerAccessGranted(
+                                service.getComponentInfo().getComponentName(), false);
+                    }
+                }
+            }
+        });
+
         // If role is not in use by other associations, revoke the role.
         // Do not need to remove the system role since it was pre-granted by the system.
         if (!isRoleInUseByOtherAssociations && deviceProfile != null && !deviceProfile.equals(
-- 
2.52.0

